rclrs/src/wait/guard_condition.rs

```diff
@@ -0,0 +1,156 @@
+
+// SAFETY: The mutexes and atomic members ensure synchronized access to members, and the callback is reentrant
+unsafe impl Sync for GuardCondition {}
+
+impl GuardCondition {
+    /// Creates a new guard condition.
+    pub fn new(context: &Context, callback: Option<Box<dyn Fn() + Send + Sync>>) -> Arc<Self> {
+        // SAFETY: Getting a zero initialized value is always safe
+        let mut guard_condition = unsafe { rcl_get_zero_initialized_guard_condition() };
+        unsafe {
+            // SAFETY: The context must be valid, and the guard condition must be zero-initialized
+            rcl_guard_condition_init(
+                &mut guard_condition,
+                &mut *context.rcl_context_mtx.lock().unwrap(),
+                rcl_guard_condition_get_default_options(),
+            );
+        }
+
+        Arc::new(Self {
+            rcl_guard_condition: Arc::new(Mutex::new(guard_condition)),
+            callback,
+            in_use_by_wait_set: Arc::new(AtomicBool::new(false)),
+        })
+    }
+
+    /// Triggers this guard condition, activating the wait set, and calling the optionally assigned callback.
+    pub fn trigger(&self) -> Result<(), RclrsError> {
+        unsafe {
+            // SAFETY: The rcl_guard_condition_t is valid.
+            rcl_trigger_guard_condition(&mut *self.rcl_guard_condition.lock().unwrap()).ok()?;
+        }
+        if let Some(callback) = &self.callback {
+            callback();
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::WaitSet;
+    use std::sync::atomic::Ordering;
+
+    #[test]
+    fn test_guard_condition() -> Result<(), RclrsError> {
+        let context = Context::new([])?;
+
+        let atomic_bool = Arc::new(std::sync::atomic::AtomicBool::new(false));
+        let atomic_bool_for_closure = Arc::clone(&atomic_bool);
+
+        let gc = GuardCondition::new(
+            &context,
+            Some(Box::new(move || {
+                atomic_bool_for_closure.store(true, Ordering::Relaxed);
+            })),
+        );
+
+        gc.trigger()?;
+
+        assert_eq!(atomic_bool.load(Ordering::Relaxed), true);
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_guard_condition_wait() -> Result<(), RclrsError> {
+        let context = Context::new([])?;
+
+        let atomic_bool = Arc::new(std::sync::atomic::AtomicBool::new(false));
+        let atomic_bool_for_closure = Arc::clone(&atomic_bool);
+
+        let gc = GuardCondition::new(
+            &context,
+            Some(Box::new(move || {
+                atomic_bool_for_closure.store(true, Ordering::Relaxed);
+            })),
+        );
+
+        let mut ws = WaitSet::new(0, 1, 0, 0, 0, 0, &context)?;
+        ws.add_guard_condition(Arc::clone(&gc))?;
+        gc.trigger()?;
+
+        assert_eq!(atomic_bool.load(Ordering::Relaxed), true);
+        ws.wait(Some(std::time::Duration::from_millis(10)))?;
+
+        Ok(())
+    }
+
+    fn assert_send<T: Send>() {}
```

Comment:
These assertions and tests are not necessary, I understand the rationale for them, but we don't test the existence of every trait we use.

Reply:
I disagree. These are not comparable to other traits like `Clone`, `Message` etc. in that they are the only auto traits, i.e.  implemented _entirely implicitly_ (apart from `Unpin`, `UnwindSafe`, `RefUnwindSafe`, which we don't care about). I would never write an assertion that `Clone` or `Message` is implemented.

Have you seen https://github.com/ros2-rust/ros2_rust/issues/253? If we accidentally make a change that makes these types not `Sync` or `Send`, we will break code.

Also, in the future, as soon as we start caring about semantic versioning, tests like these provide protection against breaking backwards compatiblity by changing auto trait implementations.

Reply:
Yes, I had seen #253, but that issue could have been detected with better tests that actually test the functionality that's provided. Testing for traits smells like bad code in my opinion, but if it's a common pattern in Rust that I'm not aware of, do you know of an example project or a documentation reference that recommends something like this?

Reply:
@esteve: https://rust-lang.github.io/api-guidelines/interoperability.html#types-are-send-and-sync-where-possible-c-send-sync

Reply:
@nnmm thanks, that was really useful, glad to be proved wrong :slightly_smiling_face: